engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java

```diff
@@ -0,0 +1,113 @@
+package org.enso.interpreter.runtime.data;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.interop.*;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import org.enso.interpreter.node.expression.builtin.interop.syntax.HostValueToEnsoNode;
+
+@ExportLibrary(InteropLibrary.class)
+public class ArraySlice implements TruffleObject {
+  private final Object storage;
+  private final long start;
+  private final long length;
+
+  public ArraySlice(Object storage, long start, long length) {
+    if (storage instanceof ArraySlice slice) {
```

Comment:
This condition is unlikely to be `true` - the code in https://github.com/enso-org/enso/pull/3724/files#r976005007 seems to wrap `ArraySlice` into yet another `Vector`.